# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/krishamerja1903/flyrank-ml-internship/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

In [13]:
from google.colab import userdata
from huggingface_hub import HfApi

HF_TOKEN = userdata.get("HF_TOKEN")

api = HfApi(token=HF_TOKEN)
user = api.whoami()

print("Hugging Face connected successfully!")
print("Logged in as:", user["name"])

Hugging Face connected successfully!
Logged in as: Krisha1903


In [14]:
from huggingface_hub import list_repo_files

files = list_repo_files(
    "FlyRank/internship-warehouse",
    repo_type="dataset",
    token=HF_TOKEN
)

for f in files:
    print(f)

.gitattributes
README.md
dim_clients.parquet
dim_content.parquet
fact_content_daily_performance/month=2025-01/data_0.parquet
fact_content_daily_performance/month=2025-02/data_0.parquet
fact_content_daily_performance/month=2025-03/data_0.parquet
fact_content_daily_performance/month=2025-04/data_0.parquet
fact_content_daily_performance/month=2025-05/data_0.parquet
fact_content_daily_performance/month=2025-06/data_0.parquet
fact_content_daily_performance/month=2025-07/data_0.parquet
fact_content_daily_performance/month=2025-08/data_0.parquet
fact_content_daily_performance/month=2025-09/data_0.parquet
fact_content_daily_performance/month=2025-10/data_0.parquet
fact_content_daily_performance/month=2025-11/data_0.parquet
fact_content_daily_performance/month=2025-12/data_0.parquet
fact_content_daily_performance/month=2026-01/data_0.parquet
fact_content_daily_performance/month=2026-02/data_0.parquet
fact_content_daily_performance/month=2026-03/data_0.parquet
fact_content_daily_performance/mont

In [15]:
from huggingface_hub import hf_hub_download
import pandas as pd

file_path = hf_hub_download(
    repo_id="FlyRank/internship-warehouse",
    filename="fact_content_daily_performance/month=2026-03/data_0.parquet",
    repo_type="dataset",
    token=HF_TOKEN
)

march = pd.read_parquet(file_path)

print("Shape:", march.shape)
print("\nColumns:")
print(march.columns.tolist())

march.head()

Shape: (9841378, 30)

Columns:
['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events']


,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,...,sessions_paid,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,True,False,True,None,20,0,67,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2026-03-01,client_73cda7b4e4f265ea,content_05597932fe4da067,True,False,True,None,1,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2026-03-01,client_73cda7b4e4f265ea,content_7a105f548d9c6916,True,False,True,None,125,1,616,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2026-03-01,client_73cda7b4e4f265ea,content_905aa32a0230694e,True,False,True,None,7,0,28,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2026-03-01,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,True,False,True,None,11,0,25,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [16]:
for i, col in enumerate(march.columns, 1):
    print(i, col)

1 report_date
2 client_hash_id
3 content_hash_id
4 client_has_gsc
5 client_has_ga4
6 gsc_data_available
7 ga4_data_available
8 gsc_impressions
9 gsc_clicks
10 gsc_sum_position
11 gsc_avg_position
12 ga4_pageviews
13 ga4_sessions
14 ga4_users
15 ga4_engaged_sessions
16 ga4_total_engagement_sec
17 sessions_organic
18 sessions_direct
19 sessions_referral
20 sessions_social
21 sessions_paid
22 sessions_ai
23 ai_chatgpt
24 ai_perplexity
25 ai_gemini
26 ai_copilot
27 ai_claude
28 ai_meta
29 ai_other
30 scroll_events


In [17]:
dim_path = hf_hub_download(
    repo_id="FlyRank/internship-warehouse",
    filename="dim_content.parquet",
    repo_type="dataset",
    token=HF_TOKEN
)

dim_content = pd.read_parquet(dim_path)

print("Shape:", dim_content.shape)

for i, col in enumerate(dim_content.columns, 1):
    print(i, col)

Shape: (519606, 26)
1 client_hash_id
2 content_hash_id
3 keyword_hash_id
4 url_hash_id
5 keyword_char_count
6 keyword_token_count
7 url_char_count
8 content_created_date
9 content_updated_date
10 content_type
11 search_volume
12 competition
13 competition_level
14 cpc
15 main_intent
16 backlinks
17 category_count
18 keyword_created_date
19 provider_used
20 model_used
21 char_count
22 word_count
23 last_optimized_date
24 optimization_eligible_date
25 is_published
26 is_deleted


In [18]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


One row represents one unique content item (web page) evaluated for refresh priority.

I will use March 2026 (month = 2026-03) as the development window.

The goal is to rank pages based on their likelihood of declining performance so the content team can decide which pages to review first.

## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

Features:
- gsc_impressions
- gsc_clicks
- gsc_avg_position
- ga4_pageviews
- ga4_engaged_sessions

Label / proxy:
- Declining performance, derived from change in historical page performance over time.

Context:
- report_date
- client_hash_id
- content_hash_id
- gsc_data_available
- ga4_data_available

Excluded:
- Future performance and any label-derived fields, because they would not be known when the refresh decision is made.

In [19]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

In [20]:
# Query 1 — Grain check
grain = (
    march.groupby(["report_date", "client_hash_id", "content_hash_id"])
    .size()
    .reset_index(name="rows_per_key")
)

print("Total rows:", len(march))
print("Unique date-client-content keys:", len(grain))
print("Maximum rows per key:", grain["rows_per_key"].max())

Total rows: 9841378
Unique date-client-content keys: 9841378
Maximum rows per key: 1


In [21]:
# Query 2 — Row count and date span
print("Row count:", len(march))
print("Start date:", march["report_date"].min())
print("End date:", march["report_date"].max())
print("Unique content items:", march["content_hash_id"].nunique())

Row count: 9841378
Start date: 2026-03-01
End date: 2026-03-31
Unique content items: 331437


In [22]:
# Query 3 — Availability check (IS TRUE equivalent)
available = march[
    march["gsc_data_available"].eq(True) &
    march["ga4_data_available"].eq(True)
]

print("Total rows:", len(march))
print("Rows with GSC and GA4 available:", len(available))
print("Rows surviving:", round(len(available) / len(march) * 100, 2), "%")

Total rows: 9841378
Rows with GSC and GA4 available: 364347
Rows surviving: 3.7 %


In [23]:
# Five-feature frame
features = available[
    [
        "gsc_impressions",
        "gsc_clicks",
        "gsc_avg_position",
        "ga4_pageviews",
        "ga4_engaged_sessions"
    ]
].copy()

print("Feature frame shape:", features.shape)
features.head()

Feature frame shape: (364347, 5)


,gsc_impressions,gsc_clicks,gsc_avg_position,ga4_pageviews,ga4_engaged_sessions
14480,5,0,5.400000,1.0,0.0
14748,39,0,5.666667,2.0,0.0
14768,179,0,5.156425,2.0,0.0
14783,72,0,7.694444,1.0,0.0
14855,3282,1,6.167885,1.0,0.0


Feature availability:

- gsc_impressions — Knowable at the decision moment because it is observed historical Search Console data.
- gsc_clicks — Knowable at the decision moment because past click counts are already observed.
- gsc_avg_position — Knowable at the decision moment because historical search position is already recorded.
- ga4_pageviews — Knowable at the decision moment because past pageviews have already occurred.
- ga4_engaged_sessions — Knowable at the decision moment because past engagement sessions are already recorded.

In [24]:
# Build a simple decline proxy from March performance
page_daily = available[
    ["report_date", "content_hash_id", "gsc_clicks"]
].copy()

page_daily["report_date"] = pd.to_datetime(page_daily["report_date"])

early = (
    page_daily[page_daily["report_date"].dt.day <= 15]
    .groupby("content_hash_id")["gsc_clicks"]
    .mean()
)

late = (
    page_daily[page_daily["report_date"].dt.day > 15]
    .groupby("content_hash_id")["gsc_clicks"]
    .mean()
)

decline = pd.concat([early, late], axis=1).dropna()
decline.columns = ["early_clicks", "late_clicks"]
decline["declining"] = (decline["late_clicks"] < decline["early_clicks"]).astype(int)

print("Pages with label:", len(decline))
print(decline["declining"].value_counts())

Pages with label: 29394
declining
0    18687
1    10707
Name: count, dtype: int64


In [25]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Aggregate the 5 honest features per page
page_features = available.groupby("content_hash_id")[
    [
        "gsc_impressions",
        "gsc_clicks",
        "gsc_avg_position",
        "ga4_pageviews",
        "ga4_engaged_sessions"
    ]
].mean()

model_data = page_features.join(decline["declining"]).dropna()

X = model_data[
    [
        "gsc_impressions",
        "gsc_clicks",
        "gsc_avg_position",
        "ga4_pageviews",
        "ga4_engaged_sessions"
    ]
]
y = model_data["declining"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

model = RandomForestClassifier(n_estimators=50, random_state=42)
model.fit(X_train, y_train)

honest_score = accuracy_score(y_test, model.predict(X_test))

# Deliberate leakage: copy the label into a feature
X_leaky = X.copy()
X_leaky["label_leak"] = y

Xl_train, Xl_test, yl_train, yl_test = train_test_split(
    X_leaky, y, test_size=0.2, random_state=42, stratify=y
)

leaky_model = RandomForestClassifier(n_estimators=50, random_state=42)
leaky_model.fit(Xl_train, yl_train)

leaky_score = accuracy_score(yl_test, leaky_model.predict(Xl_test))

print("Honest accuracy:", round(honest_score, 4))
print("Leaky accuracy:", round(leaky_score, 4))

# Remove the deliberately leaked feature
X_leaky = X_leaky.drop(columns=["label_leak"])
print("Leak removed:", "label_leak" not in X_leaky.columns)

Honest accuracy: 0.6523
Leaky accuracy: 1.0
Leak removed: True


## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

In [26]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Limitation:

Only 3.7% of March rows have both GSC and GA4 data available. Therefore, the usable slice may not represent all content equally, and the results should be treated as directional decision-support rather than a complete view of all pages.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.